File: src/engine.rs
use std::sync::Arc;
use std::sync::atomic::Ordering::Relaxed;
use std::time::Duration;

use crate::archer::{
    config::MarketConfig,
    ix_builder::{build_clear_book_ix, build_update_instructions, build_update_mid_price_ix},
};
use solana_sdk::{pubkey::Pubkey, signature::Keypair, signer::Signer};
use tokio_util::sync::CancellationToken;

use crate::{
    config::MMConfig,
    state::SharedState,
    strategy::{QuoteDecision, Strategy},
    tx::{TxPriority, TxSender},
};

const CU_CLEAR_BOOK: u32 = 650;
const CU_MID_ONLY: u32 = 850;
const CU_FULL_UPDATE: u32 = 5600;

pub async fn run_engine(
    state: Arc<SharedState>,
    sdk_config: Arc<MarketConfig>,
    mm_config: Arc<MMConfig>,
    signer: Arc<Keypair>,
    maker_pubkey: Pubkey,
    market_pubkey: Pubkey,
    tx_sender: Arc<TxSender>,
    initial_sequence_number: u64,
    is_lo: bool,
    cancel: CancellationToken,
) {
    let strategy = Strategy::new(&mm_config.strategy, is_lo);
    let heartbeat = Duration::from_millis(mm_config.execution.heartbeat_interval_ms);
    let signer_pubkey = signer.pubkey();
    let staleness_us = mm_config.feed.staleness_timeout_ms * 1000;

    let mut last_structure_hash: u64 = 0;
    let mut last_sent_mid_ticks: u64 = 0;
    let mut needs_initial_book: bool = true;
    let mut local_seq: u64 = initial_sequence_number;

    tracing::info!(
        %market_pubkey, %maker_pubkey,
        heartbeat_ms = mm_config.execution.heartbeat_interval_ms,
        num_levels = mm_config.strategy.spread_levels_bps.len(),
        book_kind = if is_lo { "LO" } else { "MM" },
        "Engine starting (event-driven + heartbeat)"
    );

    state.engine_alive.store(true, Relaxed);

    loop {
        // Wait for either a price update or the heartbeat timeout.
        let is_heartbeat = tokio::select! {
            _ = cancel.cancelled() => {
                tracing::info!("Engine shutting down, clearing book");
                state.engine_alive.store(false, Relaxed);
                local_seq += 1;
                let ix = build_clear_book_ix(&signer_pubkey, &market_pubkey, &maker_pubkey, local_seq);
                tx_sender.fire(vec![ix], TxPriority::Emergency, CU_CLEAR_BOOK);
                return;
            }
            _ = state.price_notify.notified() => false,
            _ = tokio::time::sleep(heartbeat) => true,
        };

        if cancel.is_cancelled() {
            continue; // will hit the cancel branch above
        }

        if state.consecutive_failures.load(Relaxed) >= 10 {
            local_seq += 1;
            let ix = build_clear_book_ix(&signer_pubkey, &market_pubkey, &maker_pubkey, local_seq);
            tx_sender.fire(vec![ix], TxPriority::Emergency, CU_CLEAR_BOOK);
            state.clear_book_sends.fetch_add(1, Relaxed);
            needs_initial_book = true;
            last_structure_hash = 0;
            tokio::select! {
                _ = cancel.cancelled() => return,
                _ = tokio::time::sleep(Duration::from_secs(1)) => {}
            }
            continue;
        }

        let price_age_us = crate::state::now_us().saturating_sub(state.price_timestamp_us.load(Relaxed));
        if price_age_us > staleness_us && state.price_timestamp_us.load(Relaxed) > 0 {
            tracing::warn!(age_ms = price_age_us / 1000, "Price feed stale, clearing book");
            local_seq += 1;
            let ix = build_clear_book_ix(&signer_pubkey, &market_pubkey, &maker_pubkey, local_seq);
            tx_sender.fire(vec![ix], TxPriority::Emergency, CU_CLEAR_BOOK);
            state.clear_book_sends.fetch_add(1, Relaxed);
            needs_initial_book = true;
            last_structure_hash = 0;
            continue;
        }

        let mid_price = state.mid_price.load(Relaxed);
        if mid_price <= 0.0 || !mid_price.is_finite() {
            state.cycles_total.fetch_add(1, Relaxed);
            continue;
        }

        let onchain_seq = state.onchain_sequence_number.load(Relaxed);
        if onchain_seq > local_seq {
            local_seq = onchain_seq;
        }

        let cached_mid = state.cached_mid_ticks.load(Relaxed);
        let reference_mid = if cached_mid > 0 { cached_mid } else { last_sent_mid_ticks };
        let effective_hash = if needs_initial_book { 0 } else { last_structure_hash };

        let volatility_bps = state.volatility_bps.load(Relaxed);
        let (decision, _spread_bps) = strategy.compute(
            mid_price,
            reference_mid,
            effective_hash,
            &sdk_config,
            state.base_total_lots.load(Relaxed),
            state.quote_total_lots.load(Relaxed),
            volatility_bps,
        );

        let max_dev_pct = mm_config.strategy.max_price_deviation_pct;
        if max_dev_pct > 0.0 && cached_mid > 0 {
            let candidate_mid_ticks = match &decision {
                QuoteDecision::UpdateMidOnly { new_mid_ticks } => Some(*new_mid_ticks),
                QuoteDecision::UpdateFull { book_update, .. } => Some(book_update.new_mid_price_ticks),
                _ => None,
            };
            if let Some(new_ticks) = candidate_mid_ticks {
                let dev_pct = (new_ticks as f64 - cached_mid as f64).abs() / cached_mid as f64 * 100.0;
                if dev_pct > max_dev_pct {
                    tracing::warn!(
                        new_ticks, onchain_ticks = cached_mid, dev_pct, max_dev_pct,
                        "Mid deviates beyond band — withholding update"
                    );
                    state.cycles_total.fetch_add(1, Relaxed);
                    continue;
                }
            }
        }

        match decision {
            QuoteDecision::Noop => {
                state.cycles_total.fetch_add(1, Relaxed);
                continue;
            }
            QuoteDecision::ClearBook => {
                local_seq += 1;
                let ix = build_clear_book_ix(&signer_pubkey, &market_pubkey, &maker_pubkey, local_seq);
                tx_sender.fire(vec![ix], TxPriority::Normal, CU_CLEAR_BOOK);
                state.clear_book_sends.fetch_add(1, Relaxed);
                state.updates_sent.fetch_add(1, Relaxed);
                last_structure_hash = 0;
                needs_initial_book = true;
            }
            QuoteDecision::UpdateMidOnly { new_mid_ticks } => {
                // Skip if ticks unchanged (price moved but not enough to change tick).
                if new_mid_ticks == last_sent_mid_ticks && !is_heartbeat {
                    state.cycles_total.fetch_add(1, Relaxed);
                    continue;
                }
                local_seq += 1;
                let ix = build_update_mid_price_ix(
                    &signer_pubkey, &market_pubkey, &maker_pubkey, new_mid_ticks, local_seq,
                );
                tx_sender.fire(vec![ix], TxPriority::Normal, CU_MID_ONLY);
                state.mid_only_updates.fetch_add(1, Relaxed);
                state.updates_sent.fetch_add(1, Relaxed);
                if is_heartbeat {
                    state.heartbeat_sends.fetch_add(1, Relaxed);
                }
                last_sent_mid_ticks = new_mid_ticks;
            }
            QuoteDecision::UpdateFull { ref book_update, structure_hash } => {
                local_seq += 1;
                match build_update_instructions(
                    book_update, &market_pubkey, &maker_pubkey, &signer_pubkey, local_seq,
                ) {
                    Ok(ixs) if !ixs.is_empty() => {
                        tx_sender.fire(ixs, TxPriority::Normal, CU_FULL_UPDATE);
                        state.book_updates.fetch_add(1, Relaxed);
                        state.updates_sent.fetch_add(1, Relaxed);
                        last_sent_mid_ticks = book_update.new_mid_price_ticks;
                        last_structure_hash = structure_hash;
                        needs_initial_book = false;
                    }
                    Ok(_) => { local_seq -= 1; }
                    Err(e) => {
                        tracing::warn!("build_update_instructions error: {e}");
                        local_seq -= 1;
                    }
                }
            }
        }

        state.cycles_total.fetch_add(1, Relaxed);
    }
}

File: ts/src/engine.ts
// Port of src/engine.rs — the event-driven + heartbeat quoting loop, with the
// same circuit breakers (consecutive-failure clear, staleness clear, price
// deviation band) and the same sequence-number bookkeeping.

import type { Address } from "@solana/kit";

import type { MarketConfig } from "./archer/marketConfig";
import { getClearBookInstruction } from "./generated/instructions/clearBook";
import { getUpdateBookInstruction } from "./generated/instructions/updateBook";
import { getUpdateMidPriceInstruction } from "./generated/instructions/updateMidPrice";
import type { KeyPairSigner } from "@solana/kit";
import type { MMConfig } from "./config";
import { findMakerBookPda } from "./archer/pda";
import { MAX_LEVELS, type BookUpdate } from "./archer/math";
import { nowUs, type SharedState } from "./state";
import { Strategy } from "./strategy";
import { TxPriority, type TxSender } from "./tx";

const CU_CLEAR_BOOK = 650;
const CU_MID_ONLY = 850;
const CU_FULL_UPDATE = 5600;

export interface EngineDeps {
  state: SharedState;
  sdkConfig: MarketConfig;
  mmConfig: MMConfig;
  signer: KeyPairSigner;
  makerPubkey: Address;
  marketPubkey: Address;
  txSender: TxSender;
  initialSequenceNumber: bigint;
  isLo: boolean;
  signal: AbortSignal;
}

// Pad the two ladders to MAX_LEVELS with zero levels, matching the fixed 32-slot
// UpdateBookData the program expects (bids 0..15, asks 16..31).
function buildUpdateLevels(update: BookUpdate) {
  const zero = { sizeInBaseLots: 0n, priceOffsetTicks: 0n };
  const levels = [];
  for (let i = 0; i < MAX_LEVELS; i++) {
    levels.push(update.bidLevels[i] ?? zero);
  }
  for (let i = 0; i < MAX_LEVELS; i++) {
    levels.push(update.askLevels[i] ?? zero);
  }
  return levels;
}

export async function runEngine(deps: EngineDeps): Promise<void> {
  const {
    state,
    sdkConfig,
    mmConfig,
    signer,
    makerPubkey,
    marketPubkey,
    txSender,
    initialSequenceNumber,
    isLo,
    signal,
  } = deps;

  const strategy = new Strategy(mmConfig.strategy, isLo);
  const heartbeatMs = mmConfig.execution.heartbeat_interval_ms;
  const stalenessUs = mmConfig.feed.staleness_timeout_ms * 1000;
  const [makerBookPda] = await findMakerBookPda(marketPubkey, makerPubkey);

  let lastStructureHash = 0n;
  let lastSentMidTicks = 0n;
  let needsInitialBook = true;
  let localSeq = initialSequenceNumber;

  const clearBookIx = (seq: bigint) =>
    getClearBookInstruction({
      makerAccount: signer,
      makerBookAccount: makerBookPda,
      sequenceNumber: seq,
    });

  state.engineAlive = true;

  while (true) {
    // Wait for either a price update or the heartbeat timeout.
    let isHeartbeat = false;
    let heartbeatTimer: ReturnType<typeof setTimeout> | undefined;
    const heartbeat = new Promise<void>((resolve) => {
      heartbeatTimer = setTimeout(() => {
        isHeartbeat = true;
        resolve();
      }, heartbeatMs);
    });
    const aborted = new Promise<void>((resolve) =>
      signal.addEventListener("abort", () => resolve(), { once: true }),
    );
    await Promise.race([state.priceNotify.notified(), heartbeat, aborted]);
    if (heartbeatTimer) clearTimeout(heartbeatTimer);

    if (signal.aborted) {
      console.log("Engine shutting down, clearing book");
      state.engineAlive = false;
      localSeq += 1n;
      txSender.fire([clearBookIx(localSeq)], TxPriority.Emergency, CU_CLEAR_BOOK);
      return;
    }

    if (state.consecutiveFailures >= 10) {
      localSeq += 1n;
      txSender.fire([clearBookIx(localSeq)], TxPriority.Emergency, CU_CLEAR_BOOK);
      state.clearBookSends += 1;
      needsInitialBook = true;
      lastStructureHash = 0n;
      await new Promise((r) => setTimeout(r, 1000));
      continue;
    }

    const priceAgeUs = Math.max(0, nowUs() - state.priceTimestampUs);
    if (priceAgeUs > stalenessUs && state.priceTimestampUs > 0) {
      console.warn(`Price feed stale (age ${Math.floor(priceAgeUs / 1000)}ms), clearing book`);
      localSeq += 1n;
      txSender.fire([clearBookIx(localSeq)], TxPriority.Emergency, CU_CLEAR_BOOK);
      state.clearBookSends += 1;
      needsInitialBook = true;
      lastStructureHash = 0n;
      continue;
    }

    const midPrice = state.midPrice;
    if (midPrice <= 0 || !Number.isFinite(midPrice)) {
      state.cyclesTotal += 1;
      continue;
    }

    const onchainSeq = state.onchainSequenceNumber;
    if (onchainSeq > localSeq) localSeq = onchainSeq;

    const cachedMid = state.cachedMidTicks;
    const referenceMid = cachedMid > 0n ? cachedMid : lastSentMidTicks;
    const effectiveHash = needsInitialBook ? 0n : lastStructureHash;

    const { decision } = strategy.compute(
      midPrice,
      referenceMid,
      effectiveHash,
      sdkConfig,
      state.baseTotalLots,
      state.quoteTotalLots,
      state.volatilityBps,
    );

    const maxDevPct = mmConfig.strategy.max_price_deviation_pct;
    if (maxDevPct > 0 && cachedMid > 0n) {
      let candidate: bigint | undefined;
      if (decision.kind === "updateMidOnly") candidate = decision.newMidTicks;
      else if (decision.kind === "updateFull")
        candidate = decision.bookUpdate.newMidPriceTicks;
      if (candidate !== undefined) {
        const devPct =
          (Math.abs(Number(candidate) - Number(cachedMid)) / Number(cachedMid)) * 100;
        if (devPct > maxDevPct) {
          console.warn(
            `Mid deviates beyond band (${devPct.toFixed(2)}% > ${maxDevPct}%) — withholding update`,
          );
          state.cyclesTotal += 1;
          continue;
        }
      }
    }

    switch (decision.kind) {
      case "noop":
        state.cyclesTotal += 1;
        continue;
      case "clearBook":
        localSeq += 1n;
        txSender.fire([clearBookIx(localSeq)], TxPriority.Normal, CU_CLEAR_BOOK);
        state.clearBookSends += 1;
        state.updatesSent += 1;
        lastStructureHash = 0n;
        needsInitialBook = true;
        break;
      case "updateMidOnly": {
        if (decision.newMidTicks === lastSentMidTicks && !isHeartbeat) {
          state.cyclesTotal += 1;
          continue;
        }
        localSeq += 1n;
        txSender.fire(
          [
            getUpdateMidPriceInstruction({
              makerAccount: signer,
              makerBookAccount: makerBookPda,
              newMidPriceTicks: decision.newMidTicks,
              sequenceNumber: localSeq,
            }),
          ],
          TxPriority.Normal,
          CU_MID_ONLY,
        );
        state.midOnlyUpdates += 1;
        state.updatesSent += 1;
        if (isHeartbeat) state.heartbeatSends += 1;
        lastSentMidTicks = decision.newMidTicks;
        break;
      }
      case "updateFull": {
        const update = decision.bookUpdate;
        // Mirror build_update_instructions: advance localSeq once, then the
        // prepended UpdateMidPrice (if any) uses that seq and UpdateBook uses
        // seq+1 — without advancing localSeq a second time.
        localSeq += 1n;
        let seq = localSeq;
        const ixs = [];
        if (update.midPriceChanged) {
          ixs.push(
            getUpdateMidPriceInstruction({
              makerAccount: signer,
              makerBookAccount: makerBookPda,
              newMidPriceTicks: update.newMidPriceTicks,
              sequenceNumber: seq,
            }),
          );
          seq += 1n;
        }
        ixs.push(
          getUpdateBookInstruction({
            makerAccount: signer,
            makerBookAccount: makerBookPda,
            marketAccount: marketPubkey,
            sequenceNumber: seq,
            midPriceTicks: update.newMidPriceTicks,
            numBids: Math.min(update.bidLevels.length, MAX_LEVELS),
            numAsks: Math.min(update.askLevels.length, MAX_LEVELS),
            padding: new Uint8Array(5),
            levels: buildUpdateLevels(update),
          }),
        );
        txSender.fire(ixs, TxPriority.Normal, CU_FULL_UPDATE);
        state.bookUpdates += 1;
        state.updatesSent += 1;
        lastSentMidTicks = update.newMidPriceTicks;
        lastStructureHash = decision.structureHash;
        needsInitialBook = false;
        break;
      }
    }

    state.cyclesTotal += 1;
  }
}
